src/writer.rs

```diff
@@ -59,12 +59,33 @@ use events::Event;
 pub struct Writer<W: Write> {
     /// underlying writer
     writer: W,
+    should_line_break: bool,
+    indent_char: u8,
+    indent_size: usize,
+    indents: Vec<u8>,
```

Comment:
I think it would be better to have split indentation logic out of reader logic:
```rust
struct Indentation {
    should_line_break: bool,
    indent_char: u8,
    indent_size: usize,
    indents: Vec<u8>,
}

pub struct Writer<W:Write> {
    writer: W,
    indent: Option<Indentation>,
}
```
  
and put as much logic as possible directly in impl of the struct.

Also (see my other comments) it might be a good idea to directly initialize indent with `vec![indent_char; 128]` or so.